modules/optimization/src/executioners/AdjointSolve.C

```diff
@@ -0,0 +1,163 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AdjointSolve.h"
+#include "OptimizationAppTypes.h"
+
+#include "FEProblem.h"
+#include "NonlinearSystemBase.h"
+#include "NonlinearSystem.h"
+#include "NodalBCBase.h"
+
+#include "libmesh/petsc_matrix.h"
+#include "libmesh/petsc_vector.h"
+#include "petscmat.h"
```

Comment:
In theory we shouldn't have to include these. That we do shows we need to add some functionality to the libMesh abstraction, looks like `zero_rows_and_columns`. I've created libmesh/libmesh#3540 for this

Reply:
That makes sense. I've added a request to also replicate `MatMultTranspose` which will be needed for the transient version of this.